### PR TITLE
Make JIT Serialization support arbitrary std::function<> IO

### PR DESCRIPTION
**Pull request by jjlilley**
Summary:
Right now, torch::save() uses std::ostream, which results in unnecessary
data copies in practice. Similar for torch::load().

Adding a std::function<size_t(const void*, size_t)> as an output option,
parallel to the existing filename and std::ostream apis, gives users the
flexibility to emit directly to a backing store.

For a simple case of appending the output to a std::string, we observe
significant benchmark savings (on order of -50%), even with the
minor std::function<> dispatch overhead. The main reason is that
std::ostringstream effectively requires 2 extra copies of the data
beyond a simple string.append lambda.

We also provide a parallel api for the load(), though this one is
slightly more complex due to the need to do arbitrary position reads.

### CODE REVIEW

File: caffe2/serialize/inline_container.cc [hunk 1]
```diff
@@ -232,44 +232,73 @@ PyTorchStreamReader::~PyTorchStreamReader() {
   valid("closing reader for archive ", archive_name_.c_str());
 }
 
-size_t ostream_write_func(void *pOpaque, mz_uint64 file_ofs, const void *pBuf, size_t n) {
```
Comment (resistor) [inline caffe2/serialize/inline_container.cc]: I'm not going to block on this, but it's preferable to separate unrelated formatting changes into their own diffs.

File: caffe2/serialize/inline_container.h [hunk 2]
```diff
@@ -141,18 +149,25 @@ class CAFFE2_API PyTorchStreamWriter final {
   ~PyTorchStreamWriter();
 
  private:
+  void setup(const string& file_name);
   void valid(const char* what, const char* info = "");
   size_t current_pos_ = 0;
   std::unique_ptr<mz_zip_archive> ar_;
   std::string archive_name_;
-  std::ostream* out_;
+  std::ostream* out_ = nullptr;
```
Comment (resistor) [inline caffe2/serialize/inline_container.h]: Is it really providing any value to have both paths?  Could we not just convert all callers to use the `std::function` path?

File: test/cpp/api/serialize.cpp [hunk 3]
```diff
@@ -60,6 +60,25 @@ TEST(SerializeTest, BasicToFile) {
...
+  auto x = torch::randn({5, 5});
+
+  std::string serialized;
+  torch::save(x, [&](const void* buf, size_t n) {
+    serialized.append(reinterpret_cast<const char *>(buf), n);
+    return n;
+  });
+  std::istringstream readHelper(serialized);
+  torch::Tensor y;
+  torch::load(y, readHelper);
```
Comment (resistor) [inline test/cpp/api/serialize.cpp]: For API orthogonality, we should probably make the same changes to `load` as well.

**Comment by zdevito** [comment]
This is a good idea, Ill let Owen mark when it is ready to land.

File: caffe2/serialize/inline_container.h (hunk 2 above)
Comment (jjlilley) [reply caffe2/serialize/inline_container.h]: That sounds good, I'll unify behind std::function<>.

File: test/cpp/api/serialize.cpp (hunk 3 above)
Comment (jjlilley) [reply test/cpp/api/serialize.cpp]: Yes, totally agree.
My intention was partly to gauge whether there would be objections that I hadn't thought of to this type of approach before doing the parallel load path. But I'll go ahead and change load as well.

**Comment by jjlilley** [comment]
I just uploaded an updated version:
 1) There is std::function<> support for both torch::load() and torch::save() codepaths now
 2) I did simplify/unify the lowest layer PyTorchStreamWriter impl a bit, but chose not to remove the filename-based constructor api because of the python binding to that constructor.

File: caffe2/serialize/inline_container.h [hunk 4]
```diff
@@ -123,11 +123,15 @@ class CAFFE2_API PyTorchStreamReader final {
 
 class CAFFE2_API PyTorchStreamWriter final {
  public:
-  PyTorchStreamWriter(std::string archive_name, std::ostream* out=nullptr);
-  PyTorchStreamWriter(std::ostream* out)
-  : PyTorchStreamWriter("archive", out) {}
-
-  void writeRecord(const std::string& name, const void* data, size_t size, bool compress = false);
+  explicit PyTorchStreamWriter(std::string archive_name);
```
Comment (resistor) [inline caffe2/serialize/inline_container.h]: Can't this just become a wrapper around the `std::function` constructor?

File: caffe2/serialize/inline_container.h [hunk 5]
```diff
@@ -141,13 +145,16 @@ class CAFFE2_API PyTorchStreamWriter final {
...
  private:
+  void setup(const string& file_name);
   void valid(const char* what, const char* info = "");
   size_t current_pos_ = 0;
   std::unique_ptr<mz_zip_archive> ar_;
   std::string archive_name_;
-  std::ostream* out_;
-  std::ofstream file_stream_;
+  std::string archive_name_plus_slash_;
+  std::unique_ptr<std::ofstream> file_stream_;
```
Comment (resistor) [inline caffe2/serialize/inline_container.h]: Why does this need to be a `unique_ptr`?  I beliefe `ofstream` can be default constructed.

Comment (jjlilley) [reply caffe2/serialize/inline_container.h]: It didn't need to be a unique_ptr<>, was mostly trying to save work in the case where it's not needed. Reverted this part back.

**Comment by jjlilley** [comment]
(also fixed a few issues to make clang-tidy happy)

**Comment by jjlilley** [comment]
Looked into the failing tests - related to a bug in c10/util/tempfile.h, which was constructing a temp file name with a trailing '\0' in the std::string. Uploading the fix.

**Review by resistor — APPROVED** [review]

**Comment by jjlilley** [comment]
Thanks for looking at this!

**Comment by yf225** [comment]
It seems that this PR is breaking macOS builds on master:
```
Oct 10 15:32:13 ../torch/csrc/api/src/serialize/input-archive.cpp:115:22: error: no matching function for call to 'min'
Oct 10 15:32:13       size_t nread = std::min(pos + n, size_) - pos;
Oct 10 15:32:13                      ^~~~~~~~
```

**Comment by yf225** [comment]
Sorry that to unbreak master, I will have to revert this PR.

**Comment by yf225** [comment]
We'll need to look at the Github CI on this PR - specifically `pytorch_macos_10_13_py3_build` and `pytorch_macos_10_13_cuda9_2_cudnn7_py3_build` was failing.
